Source/WebCore/dom/Document.cpp

```diff
@@ -1335,6 +1353,7 @@ void Document::setCompatibilityMode(DocumentCompatibilityMode mode)
     m_compatibilityMode = mode;
 
     if (inQuirksMode() != wasInQuirksMode) {
+        m_cachedMediaQueryDefaultStyle.reset();
```

Comment:
This doesn’t seem like sufficient invalidation. The value returned from `Style::fontSizeForKeyword` when passed `false` depends on the document’s setting’s `defaultFontSize`, so we also need to invalidate when the settings on the document are changed. While `Document::m_settings` can’t be reassigned, the values in the settings can be changed. To handle that properly, I think this needs to be invalidated, probably by `Style::Scope::didChangeStyleSheetEnvironment`. We should also consider how to test that it works properly. I suspect we can make a test that would use an internals mechanism and would demonstrate the cache doing the wrong thing.

Reply:
We could also consider naming this in a way that makes it clear that it’s a style that’s dependent only on quirks mode and font size, not more broadly on other aspects of the state of the document. I don’t immediately have a name in mind, though!

Reply:
@darinadler Thank you very much for the review! Please disregard my previous comment - it was mostly incorrect and misleading. I was tired and in a hurry yesterday, so I didn't fully understand your feedback.

Now, I believe I've managed to implement what you were suggesting.
The key realization for me was that whenever any value inside `Settings` changes, `Page::setNeedsRelayoutAllFrames` is invoked, and each document is iterated and notified about the change. That seems like the ideal place to invalidate caches that depend on document settings.
Since the code inside `Settings` is auto-generated, it appears to reliably notify about changes.